src/sender.js

```diff
@@ -20,73 +46,189 @@ const PUBLIC_KEY = {
+                break;
+            case TCP:
+                this.http = false;
+                this.secure = false;
+                break;
+            case TCPS:
+                this.http = false;
+                this.secure = true;
+                break;
+            default:
+                throw new Error(`Invalid protocol: '${options.protocol}'`);
+        }
+
+        this.host = options.host;
+        this.port = options.port;
+
+        if (this.http) {
+            this.username = options.username;
+            this.password = options.password;
+            this.token = options.token;
+        } else {
+            if (!options.auth && !options.jwk) {
+                constructAuth(options);
+            }
+            this.jwk = constructJwk(options);
+        }
+
+        this.tlsVerify = isBoolean(options.tls_verify) ? options.tls_verify : true;
+        this.tlsCA = options.tls_ca ? readFileSync(options.tls_ca) : undefined;
+
+        this.autoFlush = isBoolean(options.auto_flush) ? options.auto_flush : true;
+        this.autoFlushRows = isInteger(options.auto_flush_rows, 0) ? options.auto_flush_rows : (this.http ? DEFAULT_HTTP_AUTO_FLUSH_ROWS : DEFAULT_TCP_AUTO_FLUSH_ROWS);
+        this.autoFlushInterval = isInteger(options.auto_flush_interval, 0) ? options.auto_flush_interval : DEFAULT_AUTO_FLUSH_INTERVAL;
+
+        this.maxNameLength = isInteger(options.max_name_len, 1) ? options.max_name_len : DEFAULT_MAX_NAME_LENGTH;
+
+        this.requestMinThroughput = isInteger(options.request_min_throughput, 1) ? options.request_min_throughput : DEFAULT_REQUEST_MIN_THROUGHPUT;
+        this.requestTimeout = isInteger(options.request_timeout, 1) ? options.request_timeout : DEFAULT_REQUEST_TIMEOUT;
+        this.retryTimeout = isInteger(options.retry_timeout, 0) ? options.retry_timeout : DEFAULT_RETRY_TIMEOUT;
+
+        const noCopy = isBoolean(options.copy_buffer) && !options.copy_buffer;
         this.toBuffer = noCopy ? this.toBufferView : this.toBufferNew;
         this.doResolve = noCopy
-            ? resolve => {
-                    compact(this);
-                    resolve(true);
+            ? (resolve) => {
+                compact(this);
+                resolve(true);
+            }
+            : (resolve) => {
+                resolve(true);
             }
-            : resolve => resolve(true);
-        this.log = options && typeof options.log === 'function' && options.log ? options.log : log;
-        this.resize(options && typeof options.bufferSize === 'number' && options.bufferSize ? options.bufferSize : DEFAULT_BUFFER_SIZE);
+        this.maxBufferSize = isInteger(options.max_buf_size, 1) ? options.max_buf_size : DEFAULT_MAX_BUFFER_SIZE;
+        this.resize(isInteger(options.init_buf_size, 1) ? options.init_buf_size : DEFAULT_BUFFER_SIZE);
         this.reset();
+
+        Sender.numOfSenders++;
```

Comment:
Shouldn't we do this only for HTTP senders?

Reply:
done